subprojects/version-control/src/main/java/org/gradle/vcs/git/GitVersionControlSpec.java

```diff
@@ -35,19 +35,15 @@
      * operations which are both unsuited for simple data specification and
      * allocate additional memory.</p>
      */
-    public URI getUrl() {
-        return url;
-    }
+    URI getUrl();
 
     /**
      * Sets the URL of the repository.
      */
-    public void setUrl(URI url) {
-        this.url = url;
-    }
+    void setUrl(URI url);
 
-    @Override
-    public String getDisplayName() {
-        return "Git Repository at " + getUrl();
-    }
+    /**
+     * Sets the URL of the repository.
+     */
+    void setUrl(String url);
```

Comment:
And I guess related to this is if you want to clone from a local repo, do we force everyone to use `file://...` or do we provide a way of using `File`?

Reply:
I'm not sure about the Kotlin stuff, although, I would love to know. But, for a `File` you can just use `toURI()` on the `File` class to get an appropriate `URI`.

Reply:
Hey,

`URI` sounds good in this case. `URI`s can be relative, local or remote. Sounds like the perfect data type for what you want.

Local relative URIs would need to be "resolved" at some point tough.

In project scripts you can use `uri("relative/path/to/something")` that will create a resolved URI relative to the project dir. IIUC this piece of DSL is to be used in a settings script, which AFAIK don't have this capability. Would be nice, and probably useful for other use cases though. The project's root dir would be used as a base URI when resolving.

From the Kotlin DSL pov, you just need to have a setter accepting the very same type as the corresponding getter to get to use a nice property assignment syntax, overridden setters are then callable explicitly. Exactly like @big-guy described.

Reply:
@eskatos, I am pretty sure that the `uri()` method you are referring to also works in `settings.gradle` files. They are represented internally by: https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/initialization/SettingsScript.java which implements https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/internal/file/FileOperations.java#L37 by extending DefaultScript which extends from BasicScript which implements FileOperations.

Thanks for the advice. So, it looks like people could say either:
```
from vcs(GitVersionControlSpec) {
   url = uri("http://url")
   // or
   setUrl("http://url")
}
```

and Kotlin folks would be happy about that situation.

Reply:
👍 
Yes, that's how it's done in the Kotlin DSL, see e.g. https://github.com/gradle/kotlin-dsl/blob/develop/samples/maven-publish/build.gradle.kts#L28

Reply:
It would be interesting to see how the whole thing would be expressed in Kotlin.

Today it seems it would be something like:

```kotlin
from(
  vcs(GitVersionControlSpec::class.java) {
    url = uri("http://url")
  }
)
```

And maybe we should make it more like:

```kotlin
from { // Requires either something like a `from(Supplier<Ref>)` overload or an extension method
  vcs(GitVersionControlSpec::class) { // Requires a KClass accepting extension
    url = uri("http://url")
  }
}
```

Thoughts?

Reply:
👍 to the `from { .. }` syntax suggestion

As for `KClass/Class` mismatch, we need to figure this out holistically over the whole Gradle API, that's a Kotlin DSL concern.

Reply:
> As for KClass/Class mismatch, we need to figure this out holistically over the whole Gradle API, that's a Kotlin DSL concern.

Absolutely.

Reply:
@bamboo and @eskatos I don't understand why the `from { ... }` syntax is better? Can you explain? Also, would that have an impact on the groovy syntax as well?

Reply:
Delimiting a block with parenthesis looks awkward to me.
Yes, this was a general comment for both Groovy and Kotlin DSL.

Reply:
I had suggested it this way just because it looked a bit natural (module from vcs ...):
```
sourceControl {    
    vcsMappings {
        withModule("group:name") {
            from vcs(GitVcs) {
                url = uri("...")
            }
        }
        addRule("rule for group") { details ->
            if (details.requested.group == "group") {
                from vcs(GitVcs) {
                    url = uri("...")
                }
            }
        }
    }
}
```

And so you could share common VCS configs:
```

sourceControl {
    vcsMappings {
        def commonVcs = vcs(GitVcs) {}
        withModule("group:name") {
            from(commonVcs)
        }
        addRule("rule for group") { details ->
            if (details.requested.group == "group") {
                from(commonVcs)
            }
        }
    }
}
```

But this could be done differently by moving `vcs` directly to the mapping and ditching `from`:
```
sourceControl {    
    vcsMappings {
        withModule("group:name") {
            vcs(GitVcs) {
                url = uri("...")
            }
        }
        addRule("rule for group") { details ->
            if (details.requested.group == "group") {
                vcs(GitVcs) {
                    url = uri("...")
                }
            }
        }
    }
}
```

Is this the equivalent (different) way in Kotlin @bamboo ?
```
sourceControl {    
    vcsMappings {
        withModule("group:name") {
            vcs(GitVersionControlSpec::class.java) {
                url = uri("...")
            }
        }
        addRule("rule for group") { details ->
            if (details.requested.group == "group") {
                vcs(GitVersionControlSpec::class.java) {
                    url = uri("...")
                }
            }
        }
    }
}
```